src/OrleansRuntime/Core/Dispatcher.cs

```diff
@@ -625,23 +625,61 @@ private bool ShouldLogError(Exception ex)
                    !(ex.GetBaseException() is ClientNotAvailableException);
         }
 
-        // this is a compatibility method for portions of the code base that don't use
-        // async/await yet, which is almost everything. there's no liability to discarding the
-        // Task returned by AsyncSendMessage()
-        internal void SendMessage(Message message, ActivationData sendingActivation = null)
-        {
-            AsyncSendMessage(message, sendingActivation).Ignore();
-        }
-
-        /// <summary>
-        /// Resolve target address for a message
-        /// - use transaction info
-        /// - check ordering info in message and sending activation
-        /// - use sender's placement strategy
-        /// </summary>
-        /// <param name="message"></param>
-        /// <returns>Resolve when message is addressed (modifies message fields)</returns>
-        private async Task AddressMessage(Message message)
+		// this is a compatibility method for portions of the code base that don't use
+		// async/await yet, which is almost everything. there's no liability to discarding the
+		// Task returned by AsyncSendMessage()
+		internal void SendMessage(Message message, ActivationData sendingActivation = null)
+		{
+			if (TrySendMessageToLocalActivation(message))
```

Comment:
If I read this code correctly (and correct me please if I am wrong), this code says that when we have a message destined to a grain (no activation picked yet), we first try to see if we have a local activation for that grain and if yes, send to it. Only if not, we go via Address Msg which will lookup in the directory (or local directory cache).

If I did understand it correctly, then this is a very important and critical bug.
If we do that, it means we are not guaranteeing that eventually we will converge to a single activation (in case we had 2 activations for some time). Lets say we have 2 activations (A and B) due to silo failure, then we converged in the directory and left only B but activation A was still on silo X and silo X now due to this code keeps using A and A will forever (or for a long time) stay and serve the requests from X while the rest of the silos will be using B.

That violated our guarantees to eventually converge to 1 activation.
The solution is to always rely only on the directory which is guaranteed to converge (or the directory cache, which is also guaranteed to converge since we check it with the directory).

We had that bug, about 3-4 years ago. You will not believe how costly this bug was  and how long it took us to chase it down.